osu.Game/Rulesets/Difficulty/Utils/DecayingValue.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Diagnostics;
+
+namespace osu.Game.Rulesets.Difficulty.Utils
+{
+    /// <summary>
+    /// Represents a value that exponentially decays over time
+    /// </summary>
+    public class DecayingValue
+    {
+        private DecayingValue(double decayConstant)
+        {
+            this.decayConstant = decayConstant;
+        }
+
+        /// <param name="decayMultiplier">The value will decay by this multiplier in one second</param>
+        public static DecayingValue FromDecayMultiplierPerSecond(double decayMultiplier)
+        {
+            return new DecayingValue(Math.Log(decayMultiplier) / 1000);
+        }
+
+        /// <param name="halfLife">The value will reduced by a factor of 0.5 after this many milliseconds</param>
+        public static DecayingValue FromHalfLifeMilliseconds(double halfLife)
+        {
+            return new DecayingValue(-Math.Log(2) / halfLife);
+        }
+
+        /// <param name="decayConstant">The value will change by a factor of Math.Exp(decayConstand * elapsedTime)</param>
+        public static DecayingValue FromDecayConstant(double decayConstant)
+        {
+            return new DecayingValue(decayConstant);
+        }
```

Comment:
`FromHalfLifeMilliseconds` and `FromDecayConstant` don't appear to be used anywhere. While the extra flexibility could prove to be useful in the future, I think it's hasty to add this in from the get-go - especially since I think this is at the harm of legibility right now. See this as an example:

https://github.com/ppy/osu/blob/485fff7bac3221774a3f6bc95a2872e70691a52a/osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs#L24

I think it'd be a improvement to see the `FromDecayMultiplierPerSecond` implementation in the constructor for the time being, and consider the other methods when they become useful later.

Reply:
Disagree.

Half life is a much more intuitive way to understand exponential decay, and could be used wherever any existing decaying value is used. If I don't add it, I think it's unlikely anyone would add it in a balancing PR, but realistically I can't make use of it in this PR without making balance changes. It's a chicken/egg thing

Reply:
I'm interested in improving the code quality here over keeping decay multipliers, if others think that expressing it as a half-life is an improvement then I'd be happy to see the current implementations switch over to it. Maybe not in this particular PR though, just to keep the scope simple.